### PR TITLE
kubeadm: add v1beta4.ClusterConfiguration.EncryptionAlgorithm

### PR DESCRIPTION
**Pull request by neolit123**
#### What type of PR is this?

/kind feature
/kind deprecation

#### What this PR does / why we need it:

Add v1beta4.ClusterConfiguration.EncryptionAlgorithm field (string) and allow the user to configure the cluster asymetric encryption algorithm to be either RSA (default, 2048 pkey size) or ECDSA (P-256). Add validation and fuzzing. Conversion from v1beta3 is not required because an empty field value is accepted and defaulted to RSA if needed.

Leverage the existing configuration option (feature gate) PublicKeysECDSA but rename the backend fields, arguments, function names to be more generic - EncryptionAlgorithm instead of PublicKeyAlgorithm. That is because once the feature gate is enabled the algorithm configuration also applies to private keys. It also uses the kubeadm API type (string) instead of the x509.PublicKeyAlgorithm enum (int).

Deprecate the PublicKeysECDSA feature gate with a message. It should be removed with the release of v1beta4 or maximum one release later (it is an alpha FG).

#### Which issue(s) this PR fixes:

xref https://github.com/kubernetes/kubeadm/issues/1535

#### Special notes for your reviewer:

NONE

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by neolit123** [comment]
https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/120417/pull-kubernetes-typecheck/1698995947637313536

> ERROR(linux/arm): /home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/cmd/kube-apiserver/app/testing/testserver.go:224:4: unknown field PublicKeyAlgorithm in struct literal of type pkiutil.CertConfig
type-checking linux/amd64

when did kube-apiserver start importing kubeadm o_O.
that's a big no, no.

will send a PR a bit later.

**Comment by neolit123** [comment]
> https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/120417/pull-kubernetes-typecheck/1698995947637313536
> 
> > ERROR(linux/arm): /home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/cmd/kube-apiserver/app/testing/testserver.go:224:4: unknown field PublicKeyAlgorithm in struct literal of type pkiutil.CertConfig
> > type-checking linux/amd64
> 
> when did kube-apiserver start importing kubeadm o_O? that's a big no, no.
> 
> will send a PR a bit later. #120419

fix merged:
https://github.com/kubernetes/kubernetes/pull/120422

**Comment by neolit123** [comment]
> you can not set the release note label to "release-note-none" because the PR has the label "kind/deprecation".

/remove-kind deprecation

**Comment by neolit123** [comment]
> pull-kubernetes-e2e-gce-cos-alpha-features

https://github.com/kubernetes/kubernetes/issues/120463

**Comment by neolit123** [comment]
/retest

https://github.com/kubernetes/test-infra/pull/30633

**Comment by chendave** [comment]
/assign

will take a look in my next day.

**Comment by neolit123** [comment]
one big question was whether this option should be cluster wide. i think, yes. if per node it will be confusing to the upload-certs functionality and to the user - i.e. what algorithm is used on this node? we also lack the storage of config per node today, other that --patches.

if the user wants custom algorithm per node they must sign their own certs, i would say. it falls under advaced ux.

File: cmd/kubeadm/app/apis/kubeadm/types.go [hunk 1]
```diff
@@ -401,13 +403,18 @@ func (cfg *ClusterConfiguration) GetControlPlaneImageRepository() string {
...
 
-// PublicKeyAlgorithm returns the type of encryption keys used in the cluster.
-func (cfg *ClusterConfiguration) PublicKeyAlgorithm() x509.PublicKeyAlgorithm {
-	if features.Enabled(cfg.FeatureGates, features.PublicKeysECDSA) {
-		return x509.ECDSA
+// EncryptionAlgorithm returns the type of encryption keys used in the cluster.
+func (cfg *ClusterConfiguration) EncryptionAlgorithmType() EncryptionAlgorithmType {
+	// If the feature gate is set to true, or false respect it.
+	// If the feature gate is not set, use the EncryptionAlgorithm field (v1beta4).
+	// TODO: remove this function when the feature gate is removed.
```
Comment (chendave) [inline cmd/kubeadm/app/apis/kubeadm/types.go:410]: Do we have a timeline to remove the FG? maybe we can decide it here, saying, remove it in the next release (1.30)?

File: cmd/kubeadm/app/apis/kubeadm/v1beta3/conversion.go [hunk 2]
```diff
@@ -39,9 +39,26 @@ func Convert_v1beta3_InitConfiguration_To_kubeadm_InitConfiguration(in *InitConf
...
+	out.EncryptionAlgorithm = ""
 	return err
 }
 
+// Convert_kubeadm_ClusterConfiguration_To_v1beta3_ClusterConfiguration is an autogenerated conversion function.
+func Convert_kubeadm_ClusterConfiguration_To_v1beta3_ClusterConfiguration(in *kubeadm.ClusterConfiguration, out *ClusterConfiguration, s conversion.Scope) error {
+	return autoConvert_kubeadm_ClusterConfiguration_To_v1beta3_ClusterConfiguration(in, out, s)
+}
+
+// Convert_v1beta3_ClusterConfiguration_To_kubeadm_ClusterConfiguration is an autogenerated conversion function.
```
Comment (chendave) [inline cmd/kubeadm/app/apis/kubeadm/v1beta3/conversion.go:54]: pls remove this line, it's not "autogenerated conversion function"

File: cmd/kubeadm/app/apis/kubeadm/v1beta3/conversion.go [hunk 3]
```diff
@@ -39,9 +39,26 @@ func Convert_v1beta3_InitConfiguration_To_kubeadm_InitConfiguration(in *InitConf
...
 	}
 	err = Convert_v1beta3_ClusterConfiguration_To_kubeadm_ClusterConfiguration(&ClusterConfiguration{}, &out.ClusterConfiguration, s)
+	// Required to pass fuzzer tests. This ClusterConfiguration is empty and is never defaulted.
+	// If we call Convert_v1beta3_ClusterConfiguration_To_kubeadm_ClusterConfiguration() it will receive
+	// a default value, thus here we need to reset it back to "".
+	out.EncryptionAlgorithm = ""
 	return err
 }
 
+// Convert_kubeadm_ClusterConfiguration_To_v1beta3_ClusterConfiguration is an autogenerated conversion function.
```
Comment (chendave) [inline cmd/kubeadm/app/apis/kubeadm/v1beta3/conversion.go:49]: I think this line is not accurate enough.

File: cmd/kubeadm/app/apis/kubeadm/v1beta4/defaults.go [hunk 4]
```diff
@@ -99,6 +102,10 @@ func SetDefaults_ClusterConfiguration(obj *ClusterConfiguration) {
 		obj.ClusterName = DefaultClusterName
 	}
 
+	if obj.EncryptionAlgorithm == "" {
+		obj.EncryptionAlgorithm = DefaultEncryptionAlgorithm
+	}
+
```
Comment (chendave) [inline cmd/kubeadm/app/apis/kubeadm/v1beta4/defaults.go:108]: not necessary in this pr, we don't have testcase for xxx/defaults.go,  we might consider to improve the test coverage for this in the future.

File: cmd/kubeadm/app/apis/kubeadm/validation/validation_test.go [hunk 5]
```diff
@@ -601,7 +606,8 @@ func TestValidateInitConfiguration(t *testing.T) {
 						ServiceSubnet: "2001:db8::1/112",
 						DNSDomain:     "cluster.local",
 					},
-					CertificatesDir: "/some/other/cert/dir",
+					CertificatesDir:     "/some/other/cert/dir",
+					EncryptionAlgorithm: kubeadmapi.EncryptionAlgorithmRSA,
```
Comment (chendave) [inline cmd/kubeadm/app/apis/kubeadm/validation/validation_test.go:610]: consider to replace at least one of the algorithm here to `ECDSA`.

File: cmd/kubeadm/app/features/features.go [hunk 6]
```diff
@@ -42,7 +42,10 @@ const (
 
 // InitFeatureGates are the default feature gates for the init command
 var InitFeatureGates = FeatureList{
-	PublicKeysECDSA:      {FeatureSpec: featuregate.FeatureSpec{Default: false, PreRelease: featuregate.Alpha}},
+	PublicKeysECDSA: {
+		FeatureSpec:        featuregate.FeatureSpec{Default: false, PreRelease: featuregate.Deprecated},
+		DeprecationMessage: "The PublicKeysECDSA feature gate is deprecated and will be removed after the feature 'ClusterConfiguration.EncryptionAlgorithm' is added.",
```
Comment (chendave) [inline cmd/kubeadm/app/features/features.go:47]: okay, 'ClusterConfiguration.EncryptionAlgorithm' is added right after the merge of this pr, might be better to say exactly release cycle instead.

File: cmd/kubeadm/app/util/pkiutil/pki_helpers.go [hunk 7]
```diff
@@ -623,8 +623,8 @@ func EncodePublicKeyPEM(key crypto.PublicKey) ([]byte, error) {
 // NewPrivateKey returns a new private key.
 var NewPrivateKey = GeneratePrivateKey
 
-func GeneratePrivateKey(keyType x509.PublicKeyAlgorithm) (crypto.Signer, error) {
-	if keyType == x509.ECDSA {
+func GeneratePrivateKey(keyType kubeadmapi.EncryptionAlgorithmType) (crypto.Signer, error) {
+	if keyType == kubeadmapi.EncryptionAlgorithmECDSA {
 		return ecdsa.GenerateKey(elliptic.P256(), cryptorand.Reader)
```
Comment (chendave) [inline cmd/kubeadm/app/util/pkiutil/pki_helpers.go:628]: Just curious, why it's P256? any reference for this?

**Review by chendave — COMMENTED** [review]
Just some minor comments, this is LGTM otherwise.

File: cmd/kubeadm/app/apis/kubeadm/types.go (hunk 1 above)
Comment (neolit123) [reply cmd/kubeadm/app/apis/kubeadm/types.go:410]: right after v1beta4 is released, i'd say. unless we get complains / feedback from users.

File: cmd/kubeadm/app/apis/kubeadm/v1beta4/defaults.go (hunk 4 above)
Comment (neolit123) [reply cmd/kubeadm/app/apis/kubeadm/v1beta4/defaults.go:108]: defaulting is always required when adding new fields.
we could have better unit testing, yes.

File: cmd/kubeadm/app/features/features.go (hunk 6 above)
Comment (neolit123) [reply cmd/kubeadm/app/features/features.go:47]: https://github.com/kubernetes/kubernetes/pull/120417#discussion_r1319364251

File: cmd/kubeadm/app/util/pkiutil/pki_helpers.go (hunk 7 above)
Comment (neolit123) [reply cmd/kubeadm/app/util/pkiutil/pki_helpers.go:628]: it was a recommendation by SIG Auth IIRC, who reviewed the KEP.
our proposal to them was to use AES-GCM, symmetric encryption of minimum key size 32 bytes.
that discussion was somewhere in k/enhancements.


P-256 provides the 128bits which is what we wanted and it's a go to option for ECDSA.

> the NIST Curve P-256 which provides 128-bits of security

https://csrc.nist.gov/csrc/media/events/workshop-on-elliptic-curve-cryptography-standards/documents/papers/session6-adalier-mehmet.pdf

File: cmd/kubeadm/app/apis/kubeadm/v1beta3/conversion.go (hunk 3 above)
Comment (neolit123) [reply cmd/kubeadm/app/apis/kubeadm/v1beta3/conversion.go:49]: added updated comment to these functions.
> ... is required due to missing EncryptionAlgorithm in v1beta3.

**Comment by neolit123** [comment]
updated:
https://github.com/kubernetes/kubernetes/compare/b82ba2523ad10eada693bf9bfb8954a470d0b1d4..51732ba9ffdbb0dea62b8a101d33717cdc3913a7

**Comment by chendave** [comment]
/lgtm

thanks @neolit123 !

**Comment by chendave** [comment]
@pacoxu @SataQiu  pls take a look when you are free, I believe this PR is in good shape and ready to go now.

**Comment by neolit123** [comment]
rebased

**Review by chendave — COMMENTED** [review]
still lgtm.

/lgtm

File: cmd/kubeadm/app/apis/kubeadm/types.go [hunk 8]
```diff
@@ -518,3 +525,13 @@ type Arg struct {
...
 }
+
+// EncryptionAlgorithmType can define an asymmetric encryption algorithm type.
+type EncryptionAlgorithmType string
+
+const (
+	// EncryptionAlgorithmECDSA defines the ECDSA encryption algorithm type.
+	EncryptionAlgorithmECDSA EncryptionAlgorithmType = "ecdsa"
+	// EncryptionAlgorithmRSA defines the RSA encryption algorithm type.
+	EncryptionAlgorithmRSA EncryptionAlgorithmType = "rsa"
```
Comment (pacoxu) [inline cmd/kubeadm/app/apis/kubeadm/types.go:536]: https://github.com/golang/go/blob/ace1494d9235be94f1325ab6e45105a446b3224c/src/crypto/x509/x509.go#L263-L268 

Is it better to use `RSA` and `ECDSA` uppercase to the same as above/the original? Not sure if this must be or not.

Comment (neolit123) [reply cmd/kubeadm/app/apis/kubeadm/types.go:536]: the k8s API docs recommend uppercase for abbreviations in field names, but i don't remember if there is rule for field values.

uppercase seems better to me actually.

Comment (neolit123) [reply cmd/kubeadm/app/apis/kubeadm/types.go:536]: i had a look at what we have at https://git.k8s.io/api and it feels like there isn't a strict rule for constant abbreviation values, so there are two types:
- `cpu` all lowercase abbreviation
- `VIP` all uppercase  abbreviation
the only rule about abbreviations is to have them in consistent case e.g. don't do `Vip`, must be `vip` or `VIP`.

https://github.com/kubernetes/community/blob/ce04e02575b252d6d7baa749b74c287fd7d4a5a2/contributors/devel/sig-architecture/api-conventions.md

Comment (neolit123) [reply cmd/kubeadm/app/apis/kubeadm/types.go]: updated!

Comment (pacoxu) [reply cmd/kubeadm/app/apis/kubeadm/types.go]: /lgtm

**Comment by neolit123** [comment]
/cc @SataQiu 
any comments on this, or good to go?

**Comment by neolit123** [comment]
@pacoxu and @chendave LGTMed.
let's get back to this if there are more comments later, @SataQiu 

/hold cancel

**Pull request merged**